protocol-units/bridge/move-modules/sources/atomic_bridge_initiator.move

```diff
@@ -0,0 +1,519 @@
+
+        moveth::add_minter(master_minter, signer::address_of(account));
+
+        moveth::burn(account, @atomic_bridge, bridge_transfer.amount); 
+
+        bridge_transfer.state = COMPLETED;
+
+        event::emit_event(&mut store.bridge_transfer_completed_events, BridgeTransferCompletedEvent {
+            bridge_transfer_id: bridge_transfer_id,
+            pre_image: pre_image,
+        });
+    }
+
+    public fun refund_bridge_transfer(
+        account: &signer, 
+        bridge_transfer_id: vector<u8>,
+        atomic_bridge: &signer
+    ) acquires BridgeTransferStore, BridgeConfig {
+        let config_address = borrow_global<BridgeConfig>(@atomic_bridge).bridge_module_deployer;
+        let store = borrow_global_mut<BridgeTransferStore>(config_address);
+        let idx = get_bridge_transfer_index(&store.transfers, &bridge_transfer_id);
+        let bridge_transfer = vector::borrow_mut(&mut store.transfers, idx);
+
+        assert!(bridge_transfer.state == INITIALIZED, ENOT_INITIALIZED);
+        assert!(timestamp::now_seconds() > bridge_transfer.time_lock, 2);
+        
+        let initiator_addr = bridge_transfer.originator;
+        let bridge_addr = signer::address_of(atomic_bridge);
+        let asset = moveth::metadata();
+
+        // Transfer amount of asset from atomic bridge primary fungible store to initiator's primary fungible store
+        let initiator_store = primary_fungible_store::ensure_primary_store_exists(initiator_addr, asset);
+        let bridge_store = primary_fungible_store::ensure_primary_store_exists(@atomic_bridge, asset);
+        dispatchable_fungible_asset::transfer(atomic_bridge, bridge_store, initiator_store, bridge_transfer.amount);
+
+        bridge_transfer.state = REFUNDED;
+
+        event::emit_event(&mut store.bridge_transfer_refunded_events, BridgeTransferRefundedEvent {
+            bridge_transfer_id: bridge_transfer_id,
+        });
+    }
+
+    /// Helper function to find the index of a bridge transfer in the transfers vector
+    public fun get_bridge_transfer_index(transfers: &vector<BridgeTransfer>, bridge_transfer_id: &vector<u8>): u64 {
+        let len = vector::length(transfers);
+        let i = 0;
+        while (i < len) {
+            let transfer = vector::borrow(transfers, i);
+
+            // Create a single byte vector by concatenating all components
+            let combined_bytes = vector::empty<u8>();
+            vector::append(&mut combined_bytes, bcs::to_bytes(&transfer.originator));
+            vector::append(&mut combined_bytes, transfer.recipient);
+            vector::append(&mut combined_bytes, transfer.hash_lock);
+            vector::append(&mut combined_bytes, bcs::to_bytes(&i));
+            
+            let id = aptos_hash::keccak256(combined_bytes);
+            if (id == *bridge_transfer_id) {
+                break
+            };
+            i = i + 1;
+        };
+        i - 1
+    }
```

Comment:
This seems really inefficient, I hope there is a better way to do this. Does the movelang has something like HashMap?

Reply:
There is a map yes, `table`and `smart_table`, also as long as entries are removed after completion, this should be fine. As indexing historic data will be done via events off chain. Same as on eth side.

Reply:
Good point. I implemented `smart_table` which should be more efficient, and removed this `get_bridge_transfer_index` function.